File: reference/python/caret_v4/cleanup.py
"""The `caret-cleanup/1` spec, loaded and composed.

Cleanup is the one place a Caret backend puts the user's own words in
front of a language model and asks for them back. The wording is not
written here: it lives in the repository's canonical `spec/cleanup/v1/`
as data, and this module reads it, verifies it against its manifest, and
composes the system prompt exactly the way `composed.txt` records.

That last part is the point. The Go implementation composes the same two
files with the same rules and asserts the same equality, so a change to
the prompt that only one language notices fails a test rather than
quietly shipping two different products.
"""
from __future__ import annotations

import hashlib
import json
import os
from dataclasses import dataclass, field
from pathlib import Path

SPEC_NAME = "caret-cleanup/1"

TRANSCRIPT_OPEN_TAG = "<transcript>"
TRANSCRIPT_CLOSE_TAG = "</transcript>"

#: Byte-identical to the Go implementation's constant. Changing one
#: without the other breaks composed.txt in exactly one language.
GLOSSARY_SECTION_HEADER = (
    "Glossary. Canonical spellings for terms this speaker uses. Apply an "
    "entry only when the surrounding words clearly refer to that term. "
    "Never replace an ordinary word that merely sounds like one, and never "
    "insert a term the speaker did not say."
)

#: The files manifest.json covers, in digest order.
HASHED_FILES = ("prompt.md", "glossary.json", "composed.txt")

VOCABULARY_CONTEXT = "Supplied by the client for this dictation"


class SpecError(Exception):
    """The spec is missing, unreadable, or does not match its manifest.

    Raised at startup, never mid-dictation: a backend either has a
    verified prompt before it accepts an operation, or it says so on
    /health and returns dictation unpolished.
    """


@dataclass
class GlossaryEntry:
    canonical: str
    common_misrecognitions: list[str] = field(default_factory=list)
    context: str = ""


@dataclass
class CleanupSpec:
    directory: Path
    prompt: str
    glossary: list[GlossaryEntry]
    composed: str
    digest: str

    @property
    def spec_id(self) -> str:
        """`caret-cleanup/1 <digest>` — what /health reports. It names the
        prompt without disclosing a word of it."""
        return f"{SPEC_NAME} {self.digest}" if self.digest else SPEC_NAME

    def system_prompt(self, vocabulary: list[str] | None = None) -> str:
        """The composed system prompt, with the client's per-operation
        vocabulary folded into the glossary.

        §7 requires the vocabulary to reach the cleanup glossary, not
        just the recognizer: a name heard correctly and then "corrected"
        by the polish pass is the same bug, one step later.
        """
        entries = list(self.glossary)
        if vocabulary:
            entries.extend(_vocabulary_entries(entries, vocabulary))
        section = render_glossary_section(entries)
        return f"{self.prompt}\n\n{section}" if section else self.prompt

    def polish_prompt(self, transcript: str, framing: str | None = None) -> str:
        """The whole cleanup request as one string, for a text-in text-out
        adapter with no system-message slot. A provider with a real system
        role should send the framing and `wrap_transcript` separately —
        same two halves, same order."""
        return f"{framing or self.system_prompt()}\n\n{wrap_transcript(transcript)}"


def wrap_transcript(text: str) -> str:
    """The inert-data envelope. The transcript is passed through
    unmodified: never escaped, never trimmed, never truncated. Meaning
    preservation outranks tidiness, even for a transcript that contains
    the closing tag."""
    return f"{TRANSCRIPT_OPEN_TAG}\n{text}\n{TRANSCRIPT_CLOSE_TAG}"


def render_glossary_entry(entry: GlossaryEntry) -> str:
    line = f"- {entry.canonical}"
    context = (entry.context or "").strip()
    if context:
        line += f" — {context}"
        if not line.endswith("."):
            line += "."
    misheard = [m.strip() for m in (entry.common_misrecognitions or []) if m and m.strip()]
    if misheard:
        line += " Sometimes mis-heard as: " + ", ".join(misheard) + "."
    return line


def render_glossary_section(entries: list[GlossaryEntry]) -> str:
    """The glossary appendix, or the empty string when there is nothing
    to say — an empty glossary is invisible, not a dangling header."""
    if not entries:
        return ""
    lines = [GLOSSARY_SECTION_HEADER, ""]
    lines.extend(render_glossary_entry(entry) for entry in entries)
    return "\n".join(lines)


def _vocabulary_entries(existing: list[GlossaryEntry], vocabulary: list[str]) -> list[GlossaryEntry]:
    have = {entry.canonical.lower() for entry in existing}
    out = []
    for term in vocabulary:
        key = term.lower()
        if key in have:
            continue
        have.add(key)
        out.append(GlossaryEntry(canonical=term, context=VOCABULARY_CONTEXT))
    return out


def compute_digest(hashes: dict[str, str]) -> str:
    """A short name for "these exact spec files". The order is fixed by
    HASHED_FILES so the digest never depends on dict ordering."""
    joined = "".join(f"{name}:{hashes[name]}\n" for name in HASHED_FILES)
    return hashlib.sha256(joined.encode("utf-8")).hexdigest()[:16]


def find_spec_dir() -> Path:
    """Where the canonical spec lives.

    `CARET_CLEANUP_SPEC_DIR` points a deployment at its own vendored
    copy. Otherwise walk up from this file and the working directory,
    which finds it when running from anywhere inside the repository.
    """
    override = os.environ.get("CARET_CLEANUP_SPEC_DIR", "").strip()
    if override:
        return Path(override).expanduser()
    roots = [Path(__file__).resolve().parent, Path.cwd().resolve()]
    for root in roots:
        for directory in [root, *root.parents][:9]:
            candidate = directory / "spec" / "cleanup" / "v1"
            if (candidate / "manifest.json").is_file():
                return candidate
    raise SpecError(
        "cleanup spec not found: set CARET_CLEANUP_SPEC_DIR to a spec/cleanup/v1 directory"
    )


def load_spec(directory: str | Path | None = None) -> CleanupSpec:
    """Load and verify the spec. Every failure is a `SpecError` with a
    sentence an operator can act on."""
    directory = Path(directory) if directory else find_spec_dir()
    try:
        raw = {name: (directory / name).read_bytes() for name in HASHED_FILES}
        manifest = json.loads((directory / "manifest.json").read_text(encoding="utf-8"))
    except OSError as exc:
        raise SpecError(f"cannot read the cleanup spec from {directory}: {exc}") from exc
    except json.JSONDecodeError as exc:
        raise SpecError(f"manifest.json is not valid JSON: {exc}") from exc

    hashes = {name: hashlib.sha256(data).hexdigest() for name, data in raw.items()}
    recorded = manifest.get("files", {})
    for name in HASHED_FILES:
        if recorded.get(name) != hashes[name]:
            raise SpecError(f"{name} does not match manifest.json")
    digest = compute_digest(hashes)
    if manifest.get("digest") != digest:
        raise SpecError(
            f"manifest.json digest is {manifest.get('digest')!r}, the files digest to {digest!r}"
        )

    glossary = _parse_glossary(raw["glossary.json"], directory)
    spec = CleanupSpec(
        directory=directory,
        prompt=raw["prompt.md"].decode("utf-8").strip(),
        glossary=glossary,
        composed=raw["composed.txt"].decode("utf-8").rstrip("\n"),
        digest=digest,
    )
    # The check that catches a hand-edited prompt, and the one that keeps
    # two languages honest about each other.
    if spec.system_prompt() != spec.composed:
        raise SpecError("composed.txt is stale: prompt.md and glossary.json no longer compose to it")
    return spec


def _parse_glossary(data: bytes, directory: Path) -> list[GlossaryEntry]:
    try:
        parsed = json.loads(data.decode("utf-8"))
    except (UnicodeDecodeError, json.JSONDecodeError) as exc:
        raise SpecError(f"{directory / 'glossary.json'} is not valid JSON: {exc}") from exc
    entries = parsed.get("entries") if isinstance(parsed, dict) else parsed
    if not isinstance(entries, list):
        raise SpecError("glossary.json: expected an object with an 'entries' list")
    out = []
    for entry in entries:
        if not isinstance(entry, dict):
            raise SpecError("glossary.json: every entry must be an object")
        canonical = str(entry.get("canonical", "")).strip()
        if not canonical:
            raise SpecError("glossary.json: every entry needs a 'canonical' term")
        misheard = entry.get("common_misrecognitions") or []
        if not isinstance(misheard, list):
            raise SpecError(f"glossary.json: 'common_misrecognitions' must be a list ({canonical})")
        out.append(GlossaryEntry(
            canonical=canonical,
            common_misrecognitions=[str(m).strip() for m in misheard if str(m).strip()],
            context=str(entry.get("context", "")).strip(),
        ))
    return out

File: reference/go/cleanup.go
package caretv4

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

// Transcript cleanup, consumed as data.
//
// The wording is not written here and must never be forked into here.
// It lives in this repository's canonical spec directory,
// `spec/cleanup/v1/`, and this file reads it, checks it against its own
// manifest, and composes the system prompt the same way every other
// implementation does. `composed.txt` is the anti-drift artifact: two
// codebases that both reproduce those bytes from `prompt.md` plus the
// default glossary agree on the spec without importing each other.

// Spec identifiers.
const (
	CleanupSpecID      = "caret-cleanup"
	CleanupSpecVersion = 1
	CleanupSpecName    = "caret-cleanup/1"
)

// TranscriptOpenTag and TranscriptCloseTag delimit the inert-data
// envelope. The transcript inside is passed through unmodified: never
// escaped, never trimmed, never truncated. Meaning preservation
// outranks tidiness.
const (
	TranscriptOpenTag  = "<transcript>"
	TranscriptCloseTag = "</transcript>"
)

// glossarySectionHeader is part of the composed prompt's bytes. Changing
// a character here breaks the composed.txt check, which is the point.
const glossarySectionHeader = "Glossary. Canonical spellings for terms this speaker uses. Apply an " +
	"entry only when the surrounding words clearly refer to that term. " +
	"Never replace an ordinary word that merely sounds like one, and never " +
	"insert a term the speaker did not say."

// hashedSpecFiles are the files manifest.json covers, in digest order.
var hashedSpecFiles = []string{"prompt.md", "glossary.json", "composed.txt"}

// GlossaryEntry is one canonical spelling and the ways speech-to-text
// tends to mangle it. It biases spelling in context; it is never a
// find-and-replace.
type GlossaryEntry struct {
	Canonical             string   `json:"canonical"`
	CommonMisrecognitions []string `json:"common_misrecognitions"`
	Context               string   `json:"context"`
}

// CleanupSpec is the loaded, verified spec.
type CleanupSpec struct {
	Dir      string
	Prompt   string
	Glossary []GlossaryEntry
	Composed string
	Digest   string
}

// FindSpecDir locates `spec/cleanup/v1`. CARET_CLEANUP_SPEC_DIR points a
// deployment at its own vendored copy; otherwise the directory is found
// by walking up from the working directory and from the executable, so
// `go run`, `go test`, and an installed binary beside the repository all
// find the same bytes.
func FindSpecDir() (string, error) {
	if override := strings.TrimSpace(os.Getenv("CARET_CLEANUP_SPEC_DIR")); override != "" {
		return override, nil
	}
	var roots []string
	if wd, err := os.Getwd(); err == nil {
		roots = append(roots, wd)
	}
	if exe, err := os.Executable(); err == nil {
		roots = append(roots, filepath.Dir(exe))
	}
	for _, root := range roots {
		dir := root
		for i := 0; i < 8; i++ {
			candidate := filepath.Join(dir, "spec", "cleanup", "v1")
			if info, err := os.Stat(filepath.Join(candidate, "manifest.json")); err == nil && !info.IsDir() {
				return candidate, nil
			}
			parent := filepath.Dir(dir)
			if parent == dir {
				break
			}
			dir = parent
		}
	}
	return "", fmt.Errorf("cleanup spec not found: set CARET_CLEANUP_SPEC_DIR to a spec/cleanup/v1 directory")
}

// LoadCleanupSpec reads the spec and verifies it against its manifest.
// Every failure happens here, at configuration time, rather than in the
// middle of somebody's dictation.
func LoadCleanupSpec(dir string) (*CleanupSpec, error) {
	if dir == "" {
		found, err := FindSpecDir()
		if err != nil {
			return nil, err
		}
		dir = found
	}
	read := func(name string) (string, error) {
		b, err := os.ReadFile(filepath.Join(dir, name))
		if err != nil {
			return "", fmt.Errorf("cannot read %s: %w", name, err)
		}
		return string(b), nil
	}
	promptRaw, err := read("prompt.md")
	if err != nil {
		return nil, err
	}
	glossaryRaw, err := read("glossary.json")
	if err != nil {
		return nil, err
	}
	composedRaw, err := read("composed.txt")
	if err != nil {
		return nil, err
	}
	manifestRaw, err := read("manifest.json")
	if err != nil {
		return nil, err
	}

	var glossaryDoc struct {
		Entries []GlossaryEntry `json:"entries"`
	}
	if err := json.Unmarshal([]byte(glossaryRaw), &glossaryDoc); err != nil {
		return nil, fmt.Errorf("glossary.json is not valid JSON: %w", err)
	}
	for i, entry := range glossaryDoc.Entries {
		if strings.TrimSpace(entry.Canonical) == "" {
			return nil, fmt.Errorf("glossary.json: entry %d has no canonical term", i)
		}
	}

	var manifest struct {
		Files  map[string]string `json:"files"`
		Digest string            `json:"digest"`
	}
	if err := json.Unmarshal([]byte(manifestRaw), &manifest); err != nil {
		return nil, fmt.Errorf("manifest.json is not valid JSON: %w", err)
	}

	actual := map[string]string{
		"prompt.md":     sha256Hex([]byte(promptRaw)),
		"glossary.json": sha256Hex([]byte(glossaryRaw)),
		"composed.txt":  sha256Hex([]byte(composedRaw)),
	}
	for _, name := range hashedSpecFiles {
		if manifest.Files[name] != actual[name] {
			return nil, fmt.Errorf("%s does not match manifest.json", name)
		}
	}
	if digest := computeSpecDigest(actual); digest != manifest.Digest {
		return nil, fmt.Errorf("manifest digest is %q, files digest to %q", manifest.Digest, digest)
	}

	spec := &CleanupSpec{
		Dir:      dir,
		Prompt:   strings.TrimSpace(promptRaw),
		Glossary: glossaryDoc.Entries,
		Composed: strings.TrimRight(composedRaw, "\n"),
		Digest:   manifest.Digest,
	}
	if composed := spec.SystemPrompt(nil); composed != spec.Composed {
		return nil, fmt.Errorf("composed.txt is stale: prompt.md + glossary.json no longer compose to it")
	}
	return spec, nil
}

func sha256Hex(b []byte) string {
	sum := sha256.Sum256(b)
	return hex.EncodeToString(sum[:])
}

// computeSpecDigest is the short name for "these exact spec files":
// sha256 over "<name>:<sha256>\n" per file in fixed order, truncated to
// 16 hex characters.
func computeSpecDigest(hashes map[string]string) string {
	var joined strings.Builder
	for _, name := range hashedSpecFiles {
		fmt.Fprintf(&joined, "%s:%s\n", name, hashes[name])
	}
	return sha256Hex([]byte(joined.String()))[:16]
}

// SpecID is `caret-cleanup/1 <digest>` — safe to report over health. It
// identifies the prompt without disclosing a word of it.
func (s *CleanupSpec) SpecID() string {
	if s == nil || s.Digest == "" {
		return CleanupSpecName
	}
	return CleanupSpecName + " " + s.Digest
}

func renderGlossaryEntry(entry GlossaryEntry) string {
	line := "- " + entry.Canonical
	if ctx := strings.TrimSpace(entry.Context); ctx != "" {
		line += " — " + ctx
		if !strings.HasSuffix(line, ".") {
			line += "."
		}
	}
	var misheard []string
	for _, m := range entry.CommonMisrecognitions {
		if m = strings.TrimSpace(m); m != "" {
			misheard = append(misheard, m)
		}
	}
	if len(misheard) > 0 {
		line += " Sometimes mis-heard as: " + strings.Join(misheard, ", ") + "."
	}
	return line
}

// renderGlossarySection returns the glossary appendix, or the empty
// string when there is nothing to say, so an empty glossary is invisible
// rather than a dangling header.
func renderGlossarySection(entries []GlossaryEntry) string {
	if len(entries) == 0 {
		return ""
	}
	lines := make([]string, 0, len(entries)+2)
	lines = append(lines, glossarySectionHeader, "")
	for _, entry := range entries {
		lines = append(lines, renderGlossaryEntry(entry))
	}
	return strings.Join(lines, "\n")
}

// SystemPrompt is prompt.md plus the glossary section, with no trailing
// newline. Passing nil composes with the spec's own default glossary,
// which by construction equals composed.txt.
func (s *CleanupSpec) SystemPrompt(extra []string) string {
	entries := s.Glossary
	if len(extra) > 0 {
		entries = append(append([]GlossaryEntry{}, entries...), vocabularyEntries(entries, extra)...)
	}
	section := renderGlossarySection(entries)
	if section == "" {
		return s.Prompt
	}
	return s.Prompt + "\n\n" + section
}

// vocabularyEntries turns the operation's vocabulary list into glossary
// entries. §7 requires the list to reach the cleanup glossary, and a
// term the default glossary already carries is not repeated.
func vocabularyEntries(existing []GlossaryEntry, vocabulary []string) []GlossaryEntry {
	have := make(map[string]bool, len(existing))
	for _, entry := range existing {
		have[strings.ToLower(entry.Canonical)] = true
	}
	var out []GlossaryEntry
	for _, term := range vocabulary {
		key := strings.ToLower(term)
		if have[key] {
			continue
		}
		have[key] = true
		out = append(out, GlossaryEntry{
			Canonical: term,
			Context:   "Supplied by the client for this dictation",
		})
	}
	return out
}

// WrapTranscript is the inert-data envelope.
func WrapTranscript(text string) string {
	return TranscriptOpenTag + "\n" + text + "\n" + TranscriptCloseTag
}

// PolishPrompt is the whole cleanup request as one string, for a
// provider with no system-message slot. A provider that has one sends
// the framing as the system message and WrapTranscript as the user
// message: same two halves, same order.
func PolishPrompt(framing, transcript string) string {
	return framing + "\n\n" + WrapTranscript(transcript)
}
